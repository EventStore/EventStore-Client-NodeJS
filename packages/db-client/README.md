# @eventstore/db-client

[![npm][npm-badge]][npm-badge-url]
[![Github action CI workflow][ci-badge]][ci-badge-url]
[![license][license-badge]][license-badge-url]

This is the package for the NodeJS client for EventStoreDB 20+ and uses gRPC as the communication protocol.

## Installation

```shell script
# Yarn
$ yarn add @eventstore/db-client

# NPM
$ npm install --save @eventstore/db-client
```

## EventStoreDB Server Compatibility

This client is compatible with version `20.6.1` upwards.

Server setup instructions can be found under the installation section of the [Event Store Docs]. Follow the Docker setup for the simplest configuration.

## Example

The following snippet showcases a simple example where we form a connection, then append and read events from the server.

###### Javascript example:

```javascript
const {
  EventStoreDBClient,
  jsonEvent,
  FORWARDS,
  START,
} = require('@eventstore/db-client');

const client = new EventStoreDBClient({
  endpoint: "localhost:2113",
});

async function simpleTest() {
  const streamName = "es_supported_clients";

  const event = jsonEvent({
    type: "grpc-client",
    data: {
      languages: ["typescript", "javascript"],
      runtime: "NodeJS",
    },
  });

  const appendResult = await client.appendToStream(streamName, [event]);

    // read the event
    const events = client.readStream(streamName, {
      fromRevision: START,
      direction: FORWARDS,
      maxCount: 10,
    });

    for await (const { event } of events) {
      console.log('Appended event: ', event);
    }
  } catch (error) {
    console.error('An error occured: ', error);
  } finally {
    await client.dispose();
  }
})();
```

###### Typescript example:

```typescript
import {
  EventStoreDBClient,
  jsonEvent,
  FORWARDS,
  START,
  JSONEventType,
} from '@eventstore/db-client';

const client = new EventStoreDBClient({
  endpoint: 'localhost:2113',
});

interface Reservation {
  reservationId: string;
  movieId: string;
  userId: string;
  seatId: string;
}

type SeatReservedEvent = JSONEventType<
  'seat-reserved',
  {
    reservationId: string;
    movieId: string;
    userId: string;
    seatId: string;
  }
>;

type SeatChangedEvent = JSONEventType<
  'seat-changed',
  {
    reservationId: string;
    newSeatId: string;
  }
>;

type ReservationEvents = SeatReservedEvent | SeatChangedEvent;

async function simpleTest(): Promise<void> {
  const streamName = 'booking-abc123';

  const event = jsonEvent<SeatReservedEvent>({
    type: 'seat-reserved',
    data: {
      reservationId: 'abc123',
      movieId: 'tt0368226',
      userId: 'nm0802995',
      seatId: '4b',
    },
  });

  const appendResult = await client.appendToStream<ReservationEvents>(
    streamName,
    event
  );

  // By reading the events in the stream, we can construct the current state of the booking

  interface Reservation {
    reservationId: string;
    movieId: string;
    userId: string;
    seatId: string;
  }

  const events = client.readStream<ReservationEvents>(streamName, {
    fromRevision: START,
    direction: FORWARDS,
    maxCount: 10,
  });

  const reservation: Partial<Reservation> = {};

  for await (const { event } of events) {
    switch (event.type) {
      case 'seat-reserved': {
        reservation.reservationId = event.data.reservationId;
        reservation.movieId = event.data.movieId;
        reservation.seatId = event.data.seatId;
        reservation.userId = event.data.userId;
        break;
      }
      case 'seat-changed': {
        reservation.seatId = event.data.newSeatId;
        break;
      }
      default: {
        const _exhaustiveCheck: never = event;
        break;
      }
    }
  }
}

// Do something with our reservation
console.log(reservation);
```

## Support

Information on support can be found on our website: [Event Store Support]

## Communities

- [Discuss]
- [Discord (Event Store)][discord-event-store]
- [Discord (ddd-cqrs-es)][discord-ddd-cqrs-es]

## Contributing

Refer to our [Contribution Guidelines]

[event store support]: https://eventstore.com/support/
[discuss]: https://discuss.eventstore.com/
[discord-event-store]: https://discord.gg/Phn9pmCw3t
[discord-ddd-cqrs-es]: https://discord.com/invite/sEZGSHNNbH
[npm-badge]: https://img.shields.io/npm/v/@eventstore/db-client.svg
[npm-badge-url]: https://www.npmjs.com/package/@eventstore/db-client
[ci-badge]: https://github.com/EventStore/EventStore-Client-NodeJS/workflows/CI/badge.svg?branch=master
[ci-badge-url]: https://github.com/EventStore/EventStore-Client-NodeJS/actions
[license-badge]: https://img.shields.io/npm/l/@eventstore/db-client.svg
[license-badge-url]: https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/LICENSE
[contribution guidelines]: https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/CONTRIBUTING.md
[event store docs]: https://developers.eventstore.com/server/v20.10/docs/installation/