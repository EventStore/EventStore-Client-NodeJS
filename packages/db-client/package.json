{
  "name": "@eventstore/db-client",
  "version": "6.2.1",
  "description": "NodeJS EventStoreDB version 20+ and uses gRPC as the communication protocol.",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "npm-run-all generate build:ts",
    "build:ts": "tsc --build",
    "build:watch": "npm-run-all generate build:ts-watch",
    "build:ts-watch": "tsc --build --watch",
    "generate": "run-s generate:*",
    "generate:sed-append": "shx sed -i \"s/int64 ([A-z_]*) = ([0-9]*) \\[deprecated = true\\];/int64 \\$1 = \\$2 [deprecated = true, jstype = JS_STRING];/g\" ./protos/*",
    "generate:sed-insert": "shx sed -i \"s/int64 ([A-z_]*) = ([0-9]*);/int64 \\$1 = \\$2 [jstype = JS_STRING];/g\" ./protos/*",
    "generate:folder": "shx mkdir -p ./generated",
    "generate:js": "grpc_tools_node_protoc ./protos/*.proto --proto_path=./protos --grpc_out=grpc_js:./generated --js_out=import_style=commonjs,binary:./generated --ts_out=generate_package_definition:./generated",
    "clean": "run-s clean:*",
    "clean:generated": "shx rm -rf ./generated",
    "clean:dist": "shx rm -rf ./dist",
    "clean:buildinfo": "shx rm -rf *.tsbuildinfo"
  },
  "files": [
    "dist/",
    "generated/"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/EventStore/EventStore-Client-NodeJS.git"
  },
  "keywords": [
    "EventStore",
    "EventSourcing",
    "Client"
  ],
  "author": "Event Store Limited",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/EventStore/EventStore-Client-NodeJS/issues"
  },
  "homepage": "https://developers.eventstore.com/clients/grpc/getting-started?codeLanguage=NodeJS",
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": "^12.18.3 || >=14"
  },
  "dependencies": {
    "@grpc/grpc-js": "^1.9.12",
    "@types/debug": "^4.1.12",
    "@types/google-protobuf": "^3.15.12",
    "@types/node": "^16.18.67",
    "debug": "^4.3.2",
    "google-protobuf": "^3.21.2",
    "uuid": "^8.3.2"
  },
  "devDependencies": {
    "@types/uuid": "^8.3.4",
    "grpc-tools": "^1.12.4",
    "grpc_tools_node_protoc_ts": "^5.3.3",
    "npm-run-all": "^4.1.5",
    "shx": "^0.3.4"
  }
}
