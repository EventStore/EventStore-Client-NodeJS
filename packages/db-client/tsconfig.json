{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "module": "node16",
    "outDir": "./dist",
    "rootDir": "./src"
  },
  "include": [
    "./src/**/*.ts",
    "./generated"
  ],
  "references": [
    {
      "path": "../db-client-bridge"
    }
  ]
}
