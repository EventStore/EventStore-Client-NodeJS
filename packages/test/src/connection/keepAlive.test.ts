import { KurrentDBClient } from "@eventstore/db-client";

describe("keepAlive settings", () => {
  const endpoint = "host:1234";

  describe("should throw if < -1", () => {
    describe.each`
      option                 | value
      ${"keepAliveTimeout"}  | ${-100}
      ${"keepAliveInterval"} | ${-2}
    `("$option", ({ option, value }) => {
      test.each([
        [
          "connectionString",
          (option: string, value: number) =>
            KurrentDBClient.connectionString`esdb://host?${option}=${value}`,
        ],
        [
          "constructor",
          (option: string, value: number) =>
            new KurrentDBClient({ endpoint, [option]: value }),
        ],
      ])("%s", (_, testCase) => {
        expect(() => testCase(option, value)).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe("should warn if keepAliveInterval is less than 10_000ms (but more than -1)", () => {
    test.each([
      [
        "connectionString",
        (value: number) =>
          KurrentDBClient.connectionString`esdb://host?keepAliveInterval=${value}`,
      ],
      [
        "constructor",
        (value: number) =>
          new KurrentDBClient({ endpoint, keepAliveInterval: value }),
      ],
    ])("%s", (_, testCase) => {
      for (const keepAliveInterval of [0, 1, 10, 1000, 9999]) {
        const warnSpy = jest.spyOn(console, "warn").mockImplementation();
        testCase(keepAliveInterval);
        expect(warnSpy).toHaveBeenCalled();
        expect(warnSpy.mock.calls).toMatchSnapshot();
        warnSpy.mockRestore();
      }
    });
  });
});
