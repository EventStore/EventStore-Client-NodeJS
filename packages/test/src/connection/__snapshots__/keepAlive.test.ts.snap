// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keepAlive settings should throw if < -1 keepAliveInterval connectionString 1`] = `"Invalid keepAliveInterval of -2. Please provide a positive integer, or -1 to disable"`;

exports[`keepAlive settings should throw if < -1 keepAliveTimeout connectionString 1`] = `"Invalid keepAliveTimeout of -100. Please provide a positive integer, or -1 to disable"`;

exports[`keepAlive settings should warn if keepAliveInterval is less than 10_000ms (but more than -1) connectionString 1`] = `
[
  [
    "Specified KeepAliveInterval of 0 is less than recommended 10_000 ms.",
  ],
]
`;

exports[`keepAlive settings should warn if keepAliveInterval is less than 10_000ms (but more than -1) connectionString 2`] = `
[
  [
    "Specified KeepAliveInterval of 1 is less than recommended 10_000 ms.",
  ],
]
`;

exports[`keepAlive settings should warn if keepAliveInterval is less than 10_000ms (but more than -1) connectionString 3`] = `
[
  [
    "Specified KeepAliveInterval of 10 is less than recommended 10_000 ms.",
  ],
]
`;

exports[`keepAlive settings should warn if keepAliveInterval is less than 10_000ms (but more than -1) connectionString 4`] = `
[
  [
    "Specified KeepAliveInterval of 1000 is less than recommended 10_000 ms.",
  ],
]
`;

exports[`keepAlive settings should warn if keepAliveInterval is less than 10_000ms (but more than -1) connectionString 5`] = `
[
  [
    "Specified KeepAliveInterval of 9999 is less than recommended 10_000 ms.",
  ],
]
`;
